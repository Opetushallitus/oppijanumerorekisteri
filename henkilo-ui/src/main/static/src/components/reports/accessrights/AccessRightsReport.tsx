import React from 'react';
import { connect } from 'react-redux';
import type { L10n } from '../../../types/localisation.type';
import type { Locale } from '../../../types/locale.type';
import type { AccessRightsReportRow } from '../../../reducers/report.reducer';
import type { RootState } from '../../../store';
import { fetchAccessRightsReport, clearAccessRightsReport } from '../../../actions/report.actions';
import Loader from '../../common/icons/Loader';
import Controls from './AccessRightsReportControls';
import Report, { columns } from './AccessRightsReportData';
import exportReport from './exportUtil';

type DispatchProps = {
    fetchReport: (oid: string) => void;
    clearReport: () => void;
};

type StateProps = {
    l10n: L10n;
    locale: Locale;
    reportLoading: boolean;
    reportData?: AccessRightsReportRow[];
};

type Props = DispatchProps & StateProps;

const Header: React.FC<{ translate: (key: string) => string }> = ({ translate }) => (
    <div className="flex-horizontal">
        <span className="oph-h2 oph-bold henkilohaku-main-header">{translate('KAYTTOOIKEUSRAPORTTI_TITLE')}</span>
    </div>
);

export const AccessRightsReport: React.FC<Props> = ({
    l10n,
    locale,
    reportLoading,
    reportData,
    fetchReport,
    clearReport,
}) => {
    const [oid, setOid] = React.useState<string>(undefined);
    const [filter, setFilter] = React.useState<string>(undefined);
    const [filterValues, setFilterValues] = React.useState<string[]>([]);

    React.useEffect(() => {
        clearReport();
        oid && fetchReport(oid);
    }, [clearReport, fetchReport, oid]);

    React.useEffect(() => {
        setFilter(undefined);
        setFilterValues([...new Set((reportData || []).map((row) => row.accessRightName))]);
    }, [reportData]);

    const translate = (key: string) => l10n[locale][key] || key;

    const report = React.useMemo(
        () => reportData && reportData.filter((row) => !filter || row.accessRightName === filter),
        [reportData, filter]
    );

    return (
        <div className="wrapper">
            <Header translate={translate} />
            <Controls
                translate={translate}
                disabled={reportLoading}
                filterValues={filterValues}
                filter={filter}
                setFilter={setFilter}
                setOid={setOid}
                dataExport={report && report[0] && (() => exportReport(report, columns, translate))}
            />
            {oid && (reportLoading ? <Loader /> : <Report report={report} translate={translate} />)}
        </div>
    );
};

const mapStateToProps = (state: RootState): StateProps => ({
    l10n: state.l10n.localisations,
    locale: state.locale,
    reportLoading: state.report.reportLoading,
    reportData: state.report.reportData,
});

const mapDispatchToProps: DispatchProps = {
    fetchReport: fetchAccessRightsReport,
    clearReport: clearAccessRightsReport,
};

export default connect<StateProps, DispatchProps, {}, RootState>(
    mapStateToProps,
    mapDispatchToProps
)(AccessRightsReport);
