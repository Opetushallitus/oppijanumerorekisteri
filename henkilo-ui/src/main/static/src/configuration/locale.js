const locale = 'fi';
export default locale;