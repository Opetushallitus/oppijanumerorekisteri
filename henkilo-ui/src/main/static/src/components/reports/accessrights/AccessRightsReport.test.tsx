import React from 'react';
import { shallow } from 'enzyme';
import type { L10n } from '../../../types/localisation.type';
import { AccessRightsReport } from './AccessRightsReport';
import { Locale } from '../../../types/locale.type';

describe('AccessRightsReport', () => {
    const MINIMAL_PROPS = {
        l10n: {} as L10n,
        locale: 'fi' as Locale,
        reportLoading: false,
        reportData: undefined,
        fetchReport: jest.fn(),
        clearReport: jest.fn(),
    };

    afterEach(() => {
        jest.resetAllMocks();
    });

    afterAll(() => {
        jest.clearAllMocks();
    });

    test('renders without crashing', () => {
        expect(shallow(<AccessRightsReport {...MINIMAL_PROPS} />)).toMatchSnapshot();
    });

    test('invokes fetchReport on oid change', () => {
        // @ts-ignore
        const spy = jest.spyOn(React, 'useState').mockImplementation((_) => ['test', jest.fn()]);
        expect(MINIMAL_PROPS.fetchReport).toHaveBeenCalledTimes(0);
        shallow(<AccessRightsReport {...MINIMAL_PROPS} />);
        expect(MINIMAL_PROPS.clearReport).toHaveBeenCalled();
        expect(MINIMAL_PROPS.fetchReport).toHaveBeenCalledWith('test');
        spy.mockRestore();
    });

    test('only show report if organization selected', () => {
        expect(shallow(<AccessRightsReport {...MINIMAL_PROPS} />).find('Loader')).toHaveLength(0);
    });

    test('show spinner while loading report', () => {
        // @ts-ignore
        const spy = jest.spyOn(React, 'useState').mockImplementation((_) => ['test', jest.fn()]);
        expect(
            shallow(<AccessRightsReport {...{ ...MINIMAL_PROPS, reportLoading: true }} />).find('Loader')
        ).toHaveLength(1);
        spy.mockRestore();
    });

    test('show report when report has been loaded for given oid', () => {
        // @ts-ignore
        const spy = jest.spyOn(React, 'useState').mockImplementation((_) => ['test', jest.fn()]);
        expect(
            shallow(<AccessRightsReport {...{ ...MINIMAL_PROPS, reportData: [] }} />).find('AccessRightsReportWrapper')
        ).toHaveLength(1);
        spy.mockRestore();
    });
});
