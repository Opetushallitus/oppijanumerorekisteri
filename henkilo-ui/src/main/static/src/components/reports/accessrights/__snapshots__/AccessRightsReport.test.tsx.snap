// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessRightsReport renders without crashing 1`] = `
<div
  className="wrapper"
>
  <Header
    translate={[Function]}
  />
  <AccessRightsReportControls
    disabled={false}
    filterValues={Array []}
    setFilter={[Function]}
    setOid={[Function]}
    translate={[Function]}
  />
</div>
`;
